File: ch1/one_away.cc
// one_away checks if two strings are one or zero edits apart

#include <iostream>
#include <string>

bool add_rep(std::string &str1, std::string &str2, int len) {
  bool off = false;
  for (int i = 0; i < len; i++) {
    if (!off && str2[i] != str1[i] && str2[i] == str1[i+1]) {
      off = true;
    }
    else if (off && str2[i] != str1[i+1]) return false;
  }

  return true;
}

bool one_away(std::string &str1, std::string &str2) {
  // Get lengths of both strings
  int len1 = str1.length();
  int len2 = str2.length();

  if (len1 > len2) {
    if (len1 - len2 > 1) return false;
    return add_rep(str1, str2, len2);
  }

  else if (len2 > len1) {
    if (len2 - len1 > 1) return false;
    return add_rep(str2, str1, len1);
  }

  else {
    bool one_edit = false;
    for (int i = 0; i < len1; i++) {
      if ((str1[i] != str2[i]) && !one_edit) {
        one_edit = true;
      }

      else if ((str1[i] != str2[i]) && one_edit) {
        return false;
      }
    }

    return true;
  }
}

int main() {
  std::string str1 = "pale";
  std::string str2 = "ple";
  std::string str3 = "pales";
  std::string str4 = "bale";
  std::string str5 = "bake";
  std::cout << one_away(str1, str2) << one_away(str3, str1)
            << one_away(str1, str4) << one_away(str1, str5) << std::endl;
}

File: ch1/one_away.py
def ins_check(str1, str2, len1):
    found_diff = False
    j = 0

    # assumes str1 is the longer string
    for i in range(len1):
        if j == len1 - 1:
            break
            
        elif (str1[i] != str2[j] and not found_diff):
            found_diff = True
            continue

        elif (str1[i] != str2[j]):
            return False
        
        j += 1

    return True


def one_away(str1, str2):
    # check insert (note that inserting from one is removing from the other)
    len1 = len(str1)
    len2 = len(str2)
    
    if len1 == len2 + 1:
        return ins_check(str1, str2, len1)
    elif len2 == len1 + 1:
        return ins_check(str2, str1, len2)

    # check replace
    elif len1 == len2:
        found_diff = False
        for i in range(len1):
            if (str1[i] != str2[i] and not found_diff):
                found_diff = True
                continue

            elif (str1[i] != str2[i]):
                return False

        return True

    return False


def main():
    str1 = "pale"
    str2 = "pales"
    str3 = "ple"
    str4 = "bale"
    str5 = "bake"
    str6 = "reee"
    str7 = "ree"
    str8 = "reeee"
    str9 = "reees"

    print(one_away(str1, str3)) # T
    print(one_away(str2, str1)) # T
    print(one_away(str1, str4)) # T
    print(one_away(str1, str5)) # F
    print(one_away(str6, str7)) # T
    print(one_away(str8, str7)) # F
    print(one_away(str8, str9)) # T
    print(one_away(str7, str9)) # F
    print(one_away(str6, str9)) # T
    

if __name__ == "__main__":
    main()